matrix/src/lib.rs

```diff
@@ -160,12 +160,48 @@ pub trait Matrix<T: Send + Sync>: Send + Sync {
             .map(|r| self.padded_horizontally_packed_row(r))
     }
 
-    /// Wraps at the end.
+    /// Pack together a collection of adjacent rows from the matrix.
+    ///
+    /// Returns an iterator whose i'th element is packing of the i'th element of the
+    /// rows r through r + P::WIDTH - 1. If we exceed the height of the matrix,
+    /// wrap around and include initial rows.
     fn vertically_packed_row<P>(&self, r: usize) -> impl Iterator<Item = P>
     where
+        T: Copy,
         P: PackedValue<Value = T>,
     {
-        (0..self.width()).map(move |c| P::from_fn(|i| self.get((r + i) % self.height(), c)))
+        let rows = (0..(P::WIDTH))
+            .map(|c| self.row_slice((r + c) % self.height()))
+            .collect_vec();
+        (0..self.width()).map(move |c| P::from_fn(|i| rows[i][c]))
+    }
+
+    /// Pack together a collection of rows and "next" rows from the matrix.
+    ///
+    /// Returns a vector corresponding to 2 packed rows. The i'th element of the first
+    /// row contains the packing of the i'th element of the rows r through r + P::WIDTH - 1.
+    /// The i'th element of the second row contains the packing of the i'th element of the
+    /// rows r + step through r + step + P::WIDTH - 1. If at some point we exceed the
+    /// height of the matrix, wrap around and include initial rows.
+    ///
+    /// We assume P::WIDTH > step which lets us batch together getting the relevant matrix rows.
+    /// If a larger step size is needed, we should instead run vertically_packed_row twice.
```

Comment:
Hm is this useful because it leverages the overlap, making `rows` a bit more compact? I wouldn't have thought that'd be too important since rows shouldn't be copied (assuming `DenseMatrix` and not the default `row_slice`).

The speedup sounds very nice so wouldn't want to compromise that much, but maybe we could have two branches here to support a fallback? I think the overhead would be minimal since the branch would be highly predictable.

Reply:
Yeah it seems like you are right. I think that might have been a holdover from a previous case where I was doing some more accumulation inside `rows` but it seems with the current approach we can get away with just building two arrays.

Somewhat strangely despite `vertically_packed_row_pair` now looking identical to two copies of `vertically_packed_row` there does seem to be a small advantage in having a separate function (~1% on the examples). Probably it's making it easier for the compiler to find some optimizations or similar. I've added a comment to this effect on the file.